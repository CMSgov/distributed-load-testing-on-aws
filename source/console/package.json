{
  "name": "distributed-load-testing-on-aws-ui",
  "version": "2.0.1",
  "private": true,
  "dependencies": {
    "@aws-amplify/ui-react": "^1.0.2",
    "@fortawesome/fontawesome-svg-core": "^1.2.18",
    "@fortawesome/free-brands-svg-icons": "^5.8.2",
    "@fortawesome/free-solid-svg-icons": "^5.8.2",
    "@fortawesome/react-fontawesome": "^0.1.4",
    "aws-amplify": "^3.3.20",
    "bootstrap": "^4.6.0",
    "brace": "^0.11.1",
    "nanoid": "^3.1.25",
    "react": "^17.0.1",
    "react-ace": "^9.3.0",
    "react-dom": "^17.0.1",
    "react-router-dom": "^5.2.0",
    "react-router-hash-link": "^2.4.3",
    "react-scripts": "^4.0.3",
    "reactstrap": "^8.9.0"
  },
  "scripts": {
    "start": "react-scripts start",
    "build": "GENERATE_SOURCEMAP=false INLINE_RUNTIME_CHUNK=false react-scripts build",
    "test": "react-scripts test",
    "eject": "react-scripts eject",
    "clean": "rm -rf node_modules package-lock.json"
  },
  "eslintConfig": {
    "extends": "react-app"
  },
  "browserslist": {
    "production": [
      ">0.2%",
      "not dead",
      "not op_mini all"
    ],
    "development": [
      "last 1 chrome version",
      "last 1 firefox version",
      "last 1 safari version"
    ]
  },
  "author": "aws-solution-builders",
  "license": "Apache-2.0",
  "readme": "./README.md"
}
