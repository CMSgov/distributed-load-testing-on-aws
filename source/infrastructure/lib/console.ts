// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { CloudFrontToS3 } from '@aws-solutions-constructs/aws-cloudfront-s3';
import { Effect, PolicyStatement } from '@aws-cdk/aws-iam';
import { Bucket, IBucket } from "@aws-cdk/aws-s3";
import { Construct, Tags } from '@aws-cdk/core';
import { Function as LambdaFunction } from '@aws-cdk/aws-lambda';


/**
 * @interface DLTConsoleContructProps
 * DLTConsoleContruct props
 */
export interface DLTConsoleContructProps {
    // Custom Resource Lambda
    customResource: LambdaFunction;
    // S3 Logs Bucket
    readonly s3LogsBucket: Bucket;
    // Solution ID
    readonly solutionId: string;
}

/**
 * Distributed Load Testing on AWS console construct
 * This creates the S3 bucket and CloudFront distribution 
 * and Cognito resources for the web front end.
 * @class
 */
export class DLTConsoleContruct extends Construct {
    public cloudFrontDomainName: string;
    public consoleBucketArn: string;
    public consoleBucket: IBucket;

    constructor(scope: Construct, id: string, props: DLTConsoleContructProps) {
        super(scope, id);

        const dltS3CloudFrontDist = new CloudFrontToS3(this, 'DLTCloudFrontToS3', {
            bucketProps: {
                serverAccessLogsBucket: props.s3LogsBucket,
                serverAccessLogsPrefix: 'console-bucket-access/',
            },
            cloudFrontDistributionProps: {
                comment: 'Website distribution for the Distributed Load Testing solution',
                enableLogging: true,
                errorResponses: [
                    { httpStatus: 403, responseHttpStatus: 200, responsePagePath: '/index.html' },
                    { httpStatus: 404, responseHttpStatus: 200, responsePagePath: '/index.html' }
                ],
                httpVersion: 'http2',
                logBucket: props.s3LogsBucket,
                logFilePrefix: 'cloudfront-logs/'
            },
            insertHttpSecurityHeaders: false
        });
        Tags.of(dltS3CloudFrontDist).add('SolutionId', props.solutionId);

        this.cloudFrontDomainName = dltS3CloudFrontDist.cloudFrontWebDistribution.domainName;
        this.consoleBucket = dltS3CloudFrontDist.s3BucketInterface;

        this.consoleBucketArn = dltS3CloudFrontDist.s3BucketInterface.bucketArn;

        props.customResource.addToRolePolicy(new PolicyStatement({
            effect: Effect.ALLOW,
            actions: ['s3:PutObject'],
            resources: [this.consoleBucketArn, `${this.consoleBucketArn}/*`]
        }));
    }
}