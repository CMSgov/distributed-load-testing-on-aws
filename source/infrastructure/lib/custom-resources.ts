// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import { Aws, Construct, CustomResource } from '@aws-cdk/core';


/**
 * CustomResourcesConstruct props
 * @interface CustomResourcesConstructProps
 */
export interface CustomResourcesConstructProps {
    readonly apiEndpoint: string;
    //Custom Resource lambda
    readonly customResourceLambda: string;
    // Cognito Resources
    readonly cognitoIdentityPool: string;
    readonly cognitoUserPool: string;
    readonly cognitoUserPoolClient: string;
    // UI Console S3 bucket
    readonly consoleBucketName: string;
    // Test scenarios storage
    readonly scenariosBucket: string;
    // Source code details
    readonly sourceCodeBucketName: string;
    readonly sourceCodePrefix: string;
}

/**
 * @class
 * Distributed Load Testing on AWS Custom Resources Construct.
 * It creates a custom resource Lambda function, a solution UUID, and a custom resource to send anonymous usage.
 */
export class CustomResourcesConstruct extends Construct {

    constructor(scope: Construct, id: string, props: CustomResourcesConstructProps) {
        super(scope, id);

        new CustomResource(this, 'CopyConsoleFiles', {
            serviceToken: props.customResourceLambda,
            resourceType: 'Custom::CopyConsoleFiles',
            properties: {
                Resource: 'CopyAssets',
                SrcBucket: props.sourceCodeBucketName,
                SrcPath: `${props.sourceCodePrefix}/console`,
                ManifestFile: 'console-manifest.json',
                DestBucket: props.consoleBucketName
            }
        });

        const awsExports = `const awsConfig = {
            cw_dashboard: 'https://console.aws.amazon.com/cloudwatch/home?region=${Aws.REGION}#dashboards:name=',
            ecs_dashboard: 'https://${Aws.REGION}.console.aws.amazon.com/ecs/home?region=${Aws.REGION}#/clusters/${Aws.STACK_NAME}/tasks',
            aws_project_region: '${Aws.REGION}',
            aws_cognito_region: '${Aws.REGION}',
            aws_cognito_identity_pool_id: '${props.cognitoIdentityPool}',
            aws_user_pools_id: '${props.cognitoUserPool}',
            aws_user_pools_web_client_id: '${props.cognitoUserPoolClient}',
            oauth: {},
            aws_cloud_logic_custom: [
                {
                    name: 'dlts',
                    endpoint: '${props.apiEndpoint}',
                    region: '${Aws.REGION}'
                }
            ],
            aws_user_files_s3_bucket: '${props.scenariosBucket}',
            aws_user_files_s3_bucket_region: '${Aws.REGION}'
        }`;

        new CustomResource(this, 'ConsoleConfig', {
            serviceToken: props.customResourceLambda,
            resourceType: 'Custom::CopyConfigFiles',
            properties: {
                Resource: 'ConfigFile',
                DestBucket: props.consoleBucketName,
                AwsExports: awsExports
            }
        });
    }
}