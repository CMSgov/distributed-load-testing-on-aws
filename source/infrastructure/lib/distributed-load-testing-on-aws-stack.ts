// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import {
  Aspects,
  Aws,
  CfnCondition,
  CfnMapping,
  CfnParameter,
  CfnResource,
  Construct,
  Fn,
  IAspect,
  IConstruct,
  StackProps,
  Stack,
  CfnOutput
} from '@aws-cdk/core';
import { CognitoAuthConstruct } from './auth';
import { CommonResourcesContruct } from './common-resources';
import { FargateECSTestRunnerContruct } from './ecs';
import { FargateVpcContruct } from './vpc';
import { ScenarioTestRunnerStorageContruct } from './scenarios-storage';
import { DLTConsoleContruct } from './console';
import { CustomResourcesConstruct } from './custom-resources';
import { DLTAPI } from './api';
import { TestRunnerLambdasConstruct } from './test-task-lambdas';
import { TaskRunnerStepFunctionConstruct } from './step-functions';

/**
 * CDK Aspect implementation to set up conditions to the entire Construct resources
 */
class ConditionAspect implements IAspect {
  private readonly condition: CfnCondition;

  constructor(condition: CfnCondition) {
    this.condition = condition;
  }

  /**
   * Implement IAspect.visit to set the condition to whole resources in Construct.
   * @param {IConstruct} node Construct node to visit
   */
  visit(node: IConstruct): void {
    const resource = node as CfnResource;
    if (resource.cfnOptions) {
      resource.cfnOptions.condition = this.condition;
    }
  }
}

/**
 * Distributed Load Testing on AWS main CDK Stack
 * @class
 */
export class DLTStack extends Stack {
  // VPC ID
  private fargateVpcId: string;
  // Subnets for Fargate tasks
  private fargateSubnetA: string;
  private fargateSubnetB: string;

  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    // CFN template format version
    this.templateOptions.templateFormatVersion = '2010-09-09';

    // CFN Parameters
    // Admin name
    const adminName = new CfnParameter(this, 'AdminName', {
      type: 'String',
      description: 'Admin user name to access the Distributed Load Testing console',
      minLength: 4,
      maxLength: 20,
      allowedPattern: '[a-zA-Z0-9-]+',
      constraintDescription: "Admin username must be a minimum of 4 characters and cannot include spaces"
    });

    // Admin email
    const adminEmail = new CfnParameter(this, 'AdminEmail', {
      type: 'String',
      description: 'Admin user email address to access the Distributed Load Testing Console',
      allowedPattern: '^[_A-Za-z0-9-\\+]+(\\.[_A-Za-z0-9-]+)*@[A-Za-z0-9-]+(\\.[A-Za-z0-9]+)*(\\.[A-Za-z]{2,})$',
      constraintDescription: 'Admin email must be a valid email address',
      minLength: 5
    });

    // Existing VPC ID
    const existingVpcId = new CfnParameter(this, 'ExistingVPCId', {
      type: 'String',
      description: 'Existing VPC ID',
      allowedPattern: '(?:^$|^vpc-[a-zA-Z0-9-]+)',
    });

    const existingSubnetA = new CfnParameter(this, 'ExistingSubnetA', {
      type: 'String',
      description: 'First existing subnet',
      allowedPattern: '(?:^$|^subnet-[a-zA-Z0-9-]+)'
    });

    const existingSubnetB = new CfnParameter(this, 'ExistingSubnetB', {
      type: 'String',
      description: 'Second existing subnet',
      allowedPattern: '(?:^$|^subnet-[a-zA-Z0-9-]+)'
    });

    // VPC CIDR Block
    const vpcCidrBlock = new CfnParameter(this, 'VpcCidrBlock', {
      type: 'String',
      default: '192.168.0.0/16',
      description: 'CIDR block of the new VPC where AWS Fargate will be placed',
      allowedPattern: '(?:^$|(\\d{1,3})\\.(\\d{1,3})\\.(\\d{1,3})\\.(\\d{1,3})/(\\d{1,2}))',
      constraintDescription: 'The VPC CIDR block must be a valid IP CIDR range of the form x.x.x.x/x.',
      minLength: 9,
      maxLength: 18
    });

    // Subnet A CIDR Block
    const subnetACidrBlock = new CfnParameter(this, 'SubnetACidrBlock', {
      type: 'String',
      default: '192.168.0.0/20',
      description: 'CIDR block for subnet A of the AWS Fargate VPC',
      allowedPattern: '(?:^$|(\\d{1,3})\\.(\\d{1,3})\\.(\\d{1,3})\\.(\\d{1,3})/(\\d{1,2}))',
      constraintDescription: 'The subnet CIDR block must be a valid IP CIDR range of the form x.x.x.x/x.',
      minLength: 9,
      maxLength: 18
    });

    // Subnet B CIDR Block
    const subnetBCidrBlock = new CfnParameter(this, 'SubnetBCidrBlock', {
      type: 'String',
      default: '192.168.16.0/20',
      description: 'CIDR block for subnet B of the AWS Fargate VPC',
      allowedPattern: '(?:^$|(\\d{1,3})\\.(\\d{1,3})\\.(\\d{1,3})\\.(\\d{1,3})/(\\d{1,2}))',
      constraintDescription: 'The subnet CIDR block must be a valid IP CIDR range of the form x.x.x.x/x.'
    });

    // Egress CIDR Block
    const egressCidrBlock = new CfnParameter(this, 'EgressCidr', {
      type: 'String',
      default: '0.0.0.0/0',
      description: 'CIDR Block to restrict the ECS container outbound access',
      minLength: 9,
      maxLength: 18,
      allowedPattern: '(?:^$|(\\d{1,3})\\.(\\d{1,3})\\.(\\d{1,3})\\.(\\d{1,3})/(\\d{1,2}))',
      constraintDescription: 'The Egress CIDR block must be a valid IP CIDR range of the form x.x.x.x/x.'
    });

    // CloudFormation metadata
    this.templateOptions.metadata = {
      'AWS::CloudFormation::Interface': {
        ParameterGroups: [
          {
            Label: { default: 'Console access' },
            Parameters: [adminName.logicalId, adminEmail.logicalId]
          },
          {
            Label: { default: 'Enter values here to use your own existing VPC' },
            Parameters: [existingVpcId.logicalId, existingSubnetA.logicalId, existingSubnetB.logicalId]
          },
          {
            Label: { default: 'Or have the solution create a new AWS Fargate VPC' },
            Parameters: [vpcCidrBlock.logicalId, subnetACidrBlock.logicalId, subnetBCidrBlock.logicalId, egressCidrBlock.logicalId]
          }
        ],
        ParameterLabels: {
          [adminName.logicalId]: { default: '* Console Administrator Name' },
          [adminEmail.logicalId]: { default: '* Console Administrator Email' },
          [existingVpcId.logicalId]: { default: 'The ID of an existing VPC in this region. Ex: `vpc-1a2b3c4d5e6f`' },
          [existingSubnetA.logicalId]: { default: 'The ID of a subnet within the existing VPC. Ex: `subnet-7h8i9j0k`' },
          [existingSubnetB.logicalId]: { default: 'The ID of a subnet within the existing VPC. Ex: `subnet-1x2y3z`' },
          [vpcCidrBlock.logicalId]: { default: 'AWS Fargate VPC CIDR Block' },
          [subnetACidrBlock.logicalId]: { default: 'AWS Fargate Subnet A CIDR Block' },
          [subnetBCidrBlock.logicalId]: { default: 'AWS Fargate Subnet A CIDR Block' },
          [egressCidrBlock.logicalId]: { default: 'AWS Fargate SecurityGroup CIDR Block' }
        }
      }
    };

    // CFN Mappings
    const solutionMapping = new CfnMapping(this, 'Solution', {
      mapping: {
        Config: {
          CodeVersion: 'CODE_VERSION',
          KeyPrefix: 'SOLUTION_NAME/CODE_VERSION',
          S3Bucket: 'CODE_BUCKET',
          SendAnonymousUsage: 'Yes',
          SolutionId: 'SO0062',
          URL: 'https://metrics.awssolutionsbuilder.com/generic'
        }
      }
    });
    const sendAnonymousUsage = solutionMapping.findInMap('Config', 'SendAnonymousUsage');
    const solutionId = solutionMapping.findInMap('Config', 'SolutionId');
    const solutionVersion = solutionMapping.findInMap('Config', 'CodeVersion');
    const sourceCodeBucket = Fn.join('-', [solutionMapping.findInMap('Config', 'S3Bucket'), Aws.REGION]);
    const sourceCodePrefix = solutionMapping.findInMap('Config', 'KeyPrefix');
    const metricsUrl = solutionMapping.findInMap('Config', 'URL');

    // CFN Conditions
    const sendAnonymousUsageCondition = new CfnCondition(this, 'SendAnonymousUsage', {
      expression: Fn.conditionEquals(sendAnonymousUsage, 'Yes')
    });

    const createFargateVpcResourcesCondition = new CfnCondition(this, 'CreateFargateVPCResources', {
      expression: Fn.conditionEquals(existingVpcId.valueAsString, '')
    });

    const usingExistingVpc = new CfnCondition(this, 'BoolExistingVPC', {
      expression: Fn.conditionNot(Fn.conditionEquals(existingVpcId.valueAsString, ''))
    });

    // Fargate VPC resources  
    const fargate = new FargateVpcContruct(this, 'DLTVpc', {
      solutionId: solutionId,
      subnetACidrBlock: subnetACidrBlock.valueAsString,
      subnetBCidrBlock: subnetBCidrBlock.valueAsString,
      vpcCidrBlock: vpcCidrBlock.valueAsString,
    });
    Aspects.of(fargate).add(new ConditionAspect(createFargateVpcResourcesCondition));
    this.fargateVpcId = Fn.conditionIf(createFargateVpcResourcesCondition.logicalId,
      fargate.DLTfargateVpcId,
      existingVpcId.valueAsString
    ).toString();

    this.fargateSubnetA = Fn.conditionIf(createFargateVpcResourcesCondition.logicalId,
      fargate.subnetA,
      existingSubnetA.valueAsString
    ).toString();

    this.fargateSubnetB = Fn.conditionIf(createFargateVpcResourcesCondition.logicalId,
      fargate.subnetB,
      existingSubnetB.valueAsString
    ).toString();

    // Fargate ECS resources
    const fargateECSResources = new FargateECSTestRunnerContruct(this, 'DLTEcs', {
      DLTfargateVpcId: this.fargateVpcId,
      securityGroupEgress: egressCidrBlock.valueAsString,
      solutionId: solutionId,
    });
    const existingVpc = Fn.conditionIf(usingExistingVpc.logicalId, true, false).toString();
    const commonResources = new CommonResourcesContruct(this, 'DLTCommonResources', {
      dltEcsTaskExecutionRole: fargateECSResources.dltTaskExecutionRole,
      solutionId: solutionId,
      sourceCodeBucket,
      sourceCodePrefix,
      solutionVersion,
      sendAnonymousUsageCondition,
      existingVpc,
      metricsUrl
    });

    const dltConsole = new DLTConsoleContruct(this, 'DLTConsoleResources', {
      customResource: commonResources.customResourceLambda,
      s3LogsBucket: commonResources.s3LogsBucket,
      solutionId: solutionId
    });

    const dltStorage = new ScenarioTestRunnerStorageContruct(this, 'DLTTestRunnerStorage', {
      ecsTaskExecutionRole: fargateECSResources.dltTaskExecutionRole,
      s3LogsBucket: commonResources.s3LogsBucket,
      cloudFrontDomainName: dltConsole.cloudFrontDomainName,
      solutionId,
    });

    const stepLambdaFunctions = new TestRunnerLambdasConstruct(this, 'DLTLambdaFunction', {
      cloudWatchLogsPolicy: commonResources.cloudWatchLogsPolicy,
      dynamoDbPolicy: dltStorage.dynamoDbPolicy,
      ecsTaskExecutionRoleArn: fargateECSResources.dltTaskExecutionRole.roleArn,
      ecsCloudWatchLogGroup: fargateECSResources.dltCloudWatchLogGroup,
      ecsCluster: fargateECSResources.dltEcsClusterName,
      ecsTaskDefinition: fargateECSResources.dltTaskDefinitionArn,
      ecsTaskSecurityGroup: fargateECSResources.dltSecurityGroupId,
      scenariosS3Policy: dltStorage.scenariosS3Policy,
      subnetA: this.fargateSubnetA,
      subnetB: this.fargateSubnetB,
      metricsUrl,
      sendAnonymousUsage,
      solutionId,
      solutionVersion,
      sourceCodeBucket: commonResources.sourceBucket,
      sourceCodePrefix,
      testScenariosBucket: dltStorage.scenariosBucket.bucketName,
      testScenariosTable: dltStorage.scenariosTable,
      uuid: commonResources.uuid,
    })

    const taskRunnerStepFunctions = new TaskRunnerStepFunctionConstruct(this, 'DLTStepFunction', {
      taskStatusChecker: stepLambdaFunctions.taskStatusChecker,
      taskRunner: stepLambdaFunctions.taskRunner,
      resultsParser: stepLambdaFunctions.resultsParser,
      taskCanceler: stepLambdaFunctions.taskCanceler,
      solutionId
    });

    const dltApi = new DLTAPI(this, 'DLTApi', {
      ecsCloudWatchLogGroup: fargateECSResources.dltCloudWatchLogGroup,
      cloudWatchLogsPolicy: commonResources.cloudWatchLogsPolicy,
      dynamoDbPolicy: dltStorage.dynamoDbPolicy,
      taskCancelerInvokePolicy: stepLambdaFunctions.taskCancelerInvokePolicy,
      scenariosBucketName: dltStorage.scenariosBucket.bucketName,
      scenariosS3Policy: dltStorage.scenariosS3Policy,
      scenariosTableName: dltStorage.scenariosTable.tableName,
      ecsCuster: fargateECSResources.dltEcsClusterName,
      ecsTaskExecutionRoleArn: fargateECSResources.dltTaskExecutionRole.roleArn,
      taskRunnerStepFunctionsArn: taskRunnerStepFunctions.taskRunnerStepFunctions.stateMachineArn,
      tastCancelerArn: stepLambdaFunctions.taskCanceler.functionArn,
      metricsUrl,
      sendAnonymousUsage,
      solutionId,
      solutionVersion,
      sourceCodeBucket: commonResources.sourceBucket,
      sourceCodePrefix,
      uuid: commonResources.uuid
    });

    const cognitoResources = new CognitoAuthConstruct(this, 'DLTCognitoAuth', {
      adminEmail: adminEmail.valueAsString,
      adminName: adminName.valueAsString,
      apiId: dltApi.apiId,
      cloudFrontDomainName: dltConsole.cloudFrontDomainName,
      scenariosBucketArn: dltStorage.scenariosBucket.bucketArn,
    });

    new CustomResourcesConstruct(this, 'DLTCustomResources', {
      apiEndpoint: dltApi.apiEndpointPath,
      customResourceLambda: commonResources.customResourceLambda.functionArn,
      cognitoIdentityPool: cognitoResources.cognitoIdentityPoolId,
      cognitoUserPool: cognitoResources.cognitoUserPoolId,
      cognitoUserPoolClient: cognitoResources.cognitoUserPoolClientId,
      consoleBucketName: dltConsole.consoleBucket.bucketName,
      scenariosBucket: dltStorage.scenariosBucket.bucketName,
      sourceCodeBucketName: commonResources.sourceBucket.bucketName,
      sourceCodePrefix
    });

    //Outputs
    new CfnOutput(this, 'Console', {
      description: 'Console URL',
      value: dltConsole.cloudFrontDomainName
    });
    new CfnOutput(this, 'SolutionUUID', {
      description: 'Solution UUID',
      value: commonResources.uuid
    })
  }
}
