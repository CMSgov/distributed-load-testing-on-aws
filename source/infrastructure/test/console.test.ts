// Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import '@aws-cdk/assert/jest';
import { SynthUtils } from '@aws-cdk/assert';
import { Stack } from '@aws-cdk/core';
import { DLTConsoleContruct } from '../lib/console';
import { Code, Function as LambdaFunction, Runtime } from '@aws-cdk/aws-lambda';
import { Bucket } from '@aws-cdk/aws-s3';
import { Effect, PolicyDocument, PolicyStatement, Role, ServicePrincipal } from '@aws-cdk/aws-iam';


test('DLT API Test', () => {
  const stack = new Stack();
  const testSourceBucket = new Bucket(stack, 'testSourceCodeBucket');

  const testRole = new Role(stack, 'TestCustomResourceRole', {
    assumedBy: new ServicePrincipal('lambda.amazonaws.com'),
    inlinePolicies: {
      'DenyPolicy': new PolicyDocument({
        statements: [
          new PolicyStatement({
            effect: Effect.DENY,
            actions: ['*'],
            resources: ['*']
          })
        ]
      })
    }
  });
  const testLambda = new LambdaFunction(stack, 'TestFunction', {
    code: Code.fromBucket(Bucket.fromBucketName(stack, 'SourceCodeBucket', 'TestBucket'), 'custom-resource.zip'),
    handler: 'index.handler',
    runtime: Runtime.NODEJS_14_X,
    role: testRole
  });


  const console = new DLTConsoleContruct(stack, 'TestConsoleResources', {
    customResource: testLambda,
    s3LogsBucket: testSourceBucket,
    solutionId: 'testId',
  });

  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});
