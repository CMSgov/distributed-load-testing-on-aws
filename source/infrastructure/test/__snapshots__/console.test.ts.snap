// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DLT API Test 1`] = `
Object {
  "Resources": Object {
    "TestConsoleResourcesDLTCloudFrontToS3CloudFrontDistribution4534927E": Object {
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W70",
              "reason": "Since the distribution uses the CloudFront domain name, CloudFront automatically sets the security policy to TLSv1 regardless of the value of MinimumProtocolVersion",
            },
          ],
        },
      },
      "Properties": Object {
        "DistributionConfig": Object {
          "Comment": "Website distribution for the Distributed Load Testing solution",
          "CustomErrorResponses": Array [
            Object {
              "ErrorCode": 403,
              "ResponseCode": 200,
              "ResponsePagePath": "/index.html",
            },
            Object {
              "ErrorCode": 404,
              "ResponseCode": 200,
              "ResponsePagePath": "/index.html",
            },
          ],
          "DefaultCacheBehavior": Object {
            "CachePolicyId": "658327ea-f89d-4fab-a63d-7e88639e58f6",
            "Compress": true,
            "TargetOriginId": "TestConsoleResourcesDLTCloudFrontToS3CloudFrontDistributionOrigin1D04619FE",
            "ViewerProtocolPolicy": "redirect-to-https",
          },
          "DefaultRootObject": "index.html",
          "Enabled": true,
          "HttpVersion": "http2",
          "IPV6Enabled": true,
          "Logging": Object {
            "Bucket": Object {
              "Fn::GetAtt": Array [
                "testSourceCodeBucketC577B176",
                "RegionalDomainName",
              ],
            },
            "Prefix": "cloudfront-logs/",
          },
          "Origins": Array [
            Object {
              "DomainName": Object {
                "Fn::GetAtt": Array [
                  "TestConsoleResourcesDLTCloudFrontToS3S3BucketAC4D9E17",
                  "RegionalDomainName",
                ],
              },
              "Id": "TestConsoleResourcesDLTCloudFrontToS3CloudFrontDistributionOrigin1D04619FE",
              "S3OriginConfig": Object {
                "OriginAccessIdentity": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "origin-access-identity/cloudfront/",
                      Object {
                        "Ref": "TestConsoleResourcesDLTCloudFrontToS3CloudFrontDistributionOrigin1S3OriginDB11D6D3",
                      },
                    ],
                  ],
                },
              },
            },
          ],
        },
        "Tags": Array [
          Object {
            "Key": "SolutionId",
            "Value": "testId",
          },
        ],
      },
      "Type": "AWS::CloudFront::Distribution",
    },
    "TestConsoleResourcesDLTCloudFrontToS3CloudFrontDistributionOrigin1S3OriginDB11D6D3": Object {
      "Properties": Object {
        "CloudFrontOriginAccessIdentityConfig": Object {
          "Comment": "Identity for TestConsoleResourcesDLTCloudFrontToS3CloudFrontDistributionOrigin1D04619FE",
        },
      },
      "Type": "AWS::CloudFront::CloudFrontOriginAccessIdentity",
    },
    "TestConsoleResourcesDLTCloudFrontToS3S3BucketAC4D9E17": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "BucketEncryption": Object {
          "ServerSideEncryptionConfiguration": Array [
            Object {
              "ServerSideEncryptionByDefault": Object {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
        "LifecycleConfiguration": Object {
          "Rules": Array [
            Object {
              "NoncurrentVersionTransitions": Array [
                Object {
                  "StorageClass": "GLACIER",
                  "TransitionInDays": 90,
                },
              ],
              "Status": "Enabled",
            },
          ],
        },
        "LoggingConfiguration": Object {
          "DestinationBucketName": Object {
            "Ref": "testSourceCodeBucketC577B176",
          },
          "LogFilePrefix": "console-bucket-access/",
        },
        "PublicAccessBlockConfiguration": Object {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "Tags": Array [
          Object {
            "Key": "SolutionId",
            "Value": "testId",
          },
        ],
        "VersioningConfiguration": Object {
          "Status": "Enabled",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "TestConsoleResourcesDLTCloudFrontToS3S3BucketPolicy2004C0F3": Object {
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "F16",
              "reason": "Public website bucket policy requires a wildcard principal",
            },
          ],
        },
      },
      "Properties": Object {
        "Bucket": Object {
          "Ref": "TestConsoleResourcesDLTCloudFrontToS3S3BucketAC4D9E17",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "*",
              "Condition": Object {
                "Bool": Object {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": Object {
                "AWS": "*",
              },
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "TestConsoleResourcesDLTCloudFrontToS3S3BucketAC4D9E17",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
                Object {
                  "Fn::GetAtt": Array [
                    "TestConsoleResourcesDLTCloudFrontToS3S3BucketAC4D9E17",
                    "Arn",
                  ],
                },
              ],
              "Sid": "HttpsOnly",
            },
            Object {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Principal": Object {
                "CanonicalUser": Object {
                  "Fn::GetAtt": Array [
                    "TestConsoleResourcesDLTCloudFrontToS3CloudFrontDistributionOrigin1S3OriginDB11D6D3",
                    "S3CanonicalUserId",
                  ],
                },
              },
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    Object {
                      "Fn::GetAtt": Array [
                        "TestConsoleResourcesDLTCloudFrontToS3S3BucketAC4D9E17",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "TestCustomResourceRole4295FDB5": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": "*",
                  "Effect": "Deny",
                  "Resource": "*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "DenyPolicy",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TestCustomResourceRoleDefaultPolicy81383378": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:PutObject",
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "TestConsoleResourcesDLTCloudFrontToS3S3BucketAC4D9E17",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "TestConsoleResourcesDLTCloudFrontToS3S3BucketAC4D9E17",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TestCustomResourceRoleDefaultPolicy81383378",
        "Roles": Array [
          Object {
            "Ref": "TestCustomResourceRole4295FDB5",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TestFunction22AD90FC": Object {
      "DependsOn": Array [
        "TestCustomResourceRoleDefaultPolicy81383378",
        "TestCustomResourceRole4295FDB5",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": "TestBucket",
          "S3Key": "custom-resource.zip",
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "TestCustomResourceRole4295FDB5",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "testSourceCodeBucketC577B176": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "AccessControl": "LogDeliveryWrite",
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
